Student redirects on sign-in

Erin007 opened the pull request:
Currently, any student with progress in a course or script will be taken to the course overview page for their primary script. With this change students who are not in any sections (or are in section(s) but don't have any assigned courses/scripts) get taken to their homepage when they log in. Only students with an assigned script/course will go to the primary script overview page.

When a user signs in: 
Students with no progress and no course/script assigned => /home 
Students with no progress and assigned course/script => primary script overview page
Students with progress and no course/script assigned => /home 
Students with progress and assigned course/script => primary script overview page
All teachers => /home 

Note: If a student is assigned to a hidden script, that script will not be considered "assigned" for the purposes of this redirect.

Erin007 commented:
@ewjordan this change will likely have a merge conflict with your changes to home_controller_test regarding Clever integration. Let's keep each other in the loop about merge timelines etc. 

cc: @tanyaparker @poorvasingal

Bjvanminnen commented:
We should do some analysis of if/what sort of additional db queries this is introducing.

Bjvanminnen commented on dashboard/app/models/user.rb:
```diff
@@ -1159,26 +1159,62 @@ def in_progress_and_completed_scripts
+    assigned_scripts = []
+    assigned_user_scripts.each do |user_script|
+      assigned_scripts << Script.where(id: user_script.script_id)
+    end
```
Could be `assigned_scripts = assigned_user_scripts.map {|user_script| Script.where(id: user_script: script_id}`?

aoby replied:
Github UI won't let me add a reaction for some reason, but I +1 this!

Bjvanminnen commented on dashboard/app/models/user.rb:
```diff
@@ -1159,26 +1159,62 @@ def in_progress_and_completed_scripts
+      unless script_hidden?(script)
+        visible_assigned_scripts << script
+      end
+    end
```
i suspect this could also be simplified to something like
```
visible_assigned_scripts = assigned_scripts.select {|script| !script_hidden?(script)}
```

Bjvanminnen replied:
Why not avoid doing this work if `exclude_hidden` isn't true, i.e. something like
```
scripts = assigned_scripts
if exclude_hidden
  scripts = assigned_scripts.select {|script| !script_hidden?(script)}
end
```

aoby replied:
Also related to my earlier comment, it doesn't look like this is ever being called with `exclude_hidden` set to false. Do we even need the param and the branching logic? Can we just always exclude, or will it be included sometime later?

Bjvanminnen commented on dashboard/app/models/user.rb:
```diff
@@ -1159,26 +1159,62 @@ def in_progress_and_completed_scripts
+        link: script_path(script)
+      }
+    end
+    assigned_scripts_data
```
believe this last line is superfluous, and that you can also remove the assignment on line 1199, i.e. if the last line is just `scripts.map(...)` the result of that mapping will be what this method returns.

aoby replied:
+1 here too

aoby replied:
This `assigned_scripts` method is pretty complex. Please add unit tests and a function header.

Bjvanminnen commented on dashboard/app/models/user.rb:
```diff
@@ -1159,26 +1159,62 @@ def in_progress_and_completed_scripts
+        name: script[:name],
+        title: data_t_suffix('script.name', script[:name], 'title'),
+        description: data_t_suffix('script.name', script[:name], 'description_short', default: ''),
+        link: script_path(script)
```
This block is the same as elsewhere in the file, and I suspect we depend on the two returning identical contents. It would be good to share this code.

Bjvanminnen commented on dashboard/app/models/user.rb:
```diff
@@ -1159,26 +1159,62 @@ def in_progress_and_completed_scripts
+    assigned_courses_data
+  end
+
+  def assigned_scripts(exclude_hidden)
```
Would like to see some function level comments where possible. I realize there's not a great precedent for this in a lot of our Ruby code, but it's something I'd like to see us start getting batter at :)

aoby replied:
+1 

There is *some* precedent ;), but yeah not enough. Please follow the [yard doc format](https://www.yardoc.org/)

Bjvanminnen commented on dashboard/app/models/user.rb:
```diff
@@ -1159,26 +1159,62 @@ def in_progress_and_completed_scripts
+  end
+
+  def assigned_course_or_script?(exclude_hidden)
+    !assigned_courses.empty? || !assigned_scripts(exclude_hidden).empty?
```
Is this the only place we're using `assigned_scripts`? If so, it seems like we're doing an awful lot of work, and then just checking to see if it's empty or not. Can we get away with something simpler?

aoby replied:
It does appear to be used on line 1237 below as well. Still I agree it seems like this could potentially be simplified.

Bjvanminnen replied:
1237 is just assigned_courses (not assigned_scripts)

Erin007 replied:
True.  I don't need to know what the assigned scripts are, just if any exist. I'll simplify.

aoby commented on dashboard/app/controllers/home_controller.rb:
```diff
@@ -34,12 +34,15 @@ def health_check
 
-  # Signed in student with course progress: redirect course overview page
-  # Signed in, not student with assigned course: redirect to /home
+  # Signed in student, with an assigned course/script: redirect course overview page
```
nit: should this read "redirect *to* course overview page"?

aoby commented on dashboard/app/controllers/home_controller.rb:
```diff
@@ -34,12 +34,15 @@ def health_check
+    exclude_hidden = true
     if current_user
-      if current_user.student? && current_user.primary_script
+      if current_user.student? && current_user.assigned_course_or_script?(exclude_hidden) && current_user.primary_script
```
nit: what's the point of this `exclude_hidden` variable that is only used in this one call and never reassigned?

If you just want to call out the param name for clarity, it's better to use a keyword argument. Then you can lose the local variable and call it like `assigned_course_or_script?(exclude_hidden: true)`. 

Even better though, assuming this is the most common way that method will be called (currently it's the only way it's called), use a default value in that function's definition and only require the param when you want it to be false: 

```
def assigned_course_or_script?(exclude_hidden: true)
...
```

aoby commented on dashboard/app/models/user.rb:
```diff
@@ -1159,26 +1159,62 @@ def in_progress_and_completed_scripts
+    assigned_user_scripts = user_scripts.where("assigned_at")
+
+    # get the scripts associated with the assigned_user_scripts
+    assigned_scripts = []
```
nit, it seems a bit odd to use the same name as the method for this variable. It's probably fine though.

Erin007 commented:
Additional db queries: 
- `assigned_courses`, no new queries, this new method just extracts the queries for courses that were previously happening in `recent_courses_and_scripts`
- `any_visible_assigned_scripts?`, 2 new queries, one to find assigned user_scripts and one to find scripts associated with the assigned user_scripts 
- `assigned_course_or_script?`, queries from `assigned_courses` and `any_visible_assigned_scripts?`

Is that accurate? Reasonable? @wjordan

Erin007 commented on dashboard/app/controllers/home_controller.rb:
```diff
@@ -35,12 +35,14 @@ def health_check
   def index
     if current_user
-      if current_user.student? && current_user.primary_script && !session['clever_link_flag']
+      if current_user.student? && current_user.assigned_course_or_script? && !session['clever_link_flag']
```
It was redundant to ask check here for primary_script because it the user has an assigned_course_or_script, they will have a primary_script.

Erin007 replied:
Ugh, actually maybe not. Fixed in #18572

Bjvanminnen commented:
I would say we need signoff from @wjordan before adding any new queries to the homepage.

Erin007 commented:
This is ready for another glance when you have time.  I simplified the check for assigned scripts, added more test coverage for the new methods and incorporate the previous feedback. Thanks!

Bjvanminnen commented on dashboard/app/models/user.rb:
```diff
@@ -1159,28 +1159,48 @@ def in_progress_and_completed_scripts
         link: course_path(course),
       }
     end
+    assigned_courses_data
```
no need for `assigned_courses_data`: can just return `section_courses.map(...)`

Erin007 replied:
The linter didn't like that.

 dashboard/app/models/user.rb:1165:5: W: Lint/UselessAssignment: Useless assignment to variable - assigned_courses_data.
    assigned_courses_data = section_courses.map do |course|
    ^^^^^^^^^^^^^^^^^^^^^

Bjvanminnen replied:
you dont need `assigned_courses_data =` either

Bjvanminnen commented on dashboard/app/models/user.rb:
```diff
@@ -1159,28 +1159,48 @@ def in_progress_and_completed_scripts
 
+    assigned_visible_scripts = assigned_user_scripts.map {|user_script| Script.where(id: user_script.script_id, hidden: 'false')}
+
+    !assigned_visible_scripts.flatten.empty?
```
How I might write this:
```ruby

user_scripts.where("assigned_at").
  map {|user_script| Script.where(id: user_script.script.id, hidden: 'false')}.
  flatten.
  any?
```

Bjvanminnen commented on dashboard/app/models/user.rb:
```diff
@@ -1159,28 +1159,48 @@ def in_progress_and_completed_scripts
+  # Checks if there are any non-hidden scripts or courses assigned to the user.
+  # @returns boolean
+  def assigned_course_or_script?
+    !assigned_courses.empty? || any_visible_assigned_scripts?
```
I think you could also use `assigned_courses.any?` here instead of `!assigned_courses.empty?`. Not sure if people with more ruby knowledge like @aoby have guidance here - I think one place .any? can get you in trouble is that `[false].any?` is false (i.e. it looks for any truthy value).

aoby replied:
Agreed, `any?` is preferable to `!empty?`, and presumably assigned_courses will all be models, and therefore truthy.

Bjvanminnen commented:
Changes generally LGTM from a code perspective, but we will want Will to sign off on any PR that adds more db queries to the homepage.

aoby commented on dashboard/app/models/user.rb:
```diff
@@ -1159,28 +1159,48 @@ def in_progress_and_completed_scripts
+  # Return a collection of courses and scripts for the user.
+  # First in the list will be courses enrolled in by the user's sections.
+  # Following that will be all scripts in which the user has made progress that # are not in any of the enrolled courses.
+  def recent_courses_and_scripts(exclude_primary_script)
```
Add a `@param`  explaining `exclude_primary_script`, and a `@returns` following the [yarddoc](http://www.rubydoc.info/gems/yard/file/docs/GettingStarted.md) format.

aoby commented on dashboard/test/models/user_test.rb:
```diff
@@ -2024,6 +2024,65 @@ def mock_geocoder_result(result)
+
+    test "it checks visibility of assigned scripts" do
+      # no assigned scripts
+      assert_equal false, @student.any_visible_assigned_scripts?
```
`refute @student.any_visible_assigned_scripts?`

aoby commented on dashboard/test/models/user_test.rb:
```diff
@@ -2024,6 +2024,65 @@ def mock_geocoder_result(result)
+      # assigned hidden script
+      hidden_script = create :script, name: 'hidden-script', hidden: true
+      @student.assign_script(hidden_script)
+      assert_equal false, @student.any_visible_assigned_scripts?
```
same as above, use `refute`

aoby commented on dashboard/test/models/user_test.rb:
```diff
@@ -2024,6 +2024,65 @@ def mock_geocoder_result(result)
+      # assigned a hidden script and a visible script
+      visible_script = create :script, name: 'visible-script'
+      @student.assign_script(visible_script)
+      assert_equal true, @student.any_visible_assigned_scripts?
```
Unless you specifically want to make sure that it's exactly `true`, you can just use `assert` here (and `refute` above)

aoby commented on dashboard/test/models/user_test.rb:
```diff
@@ -2024,6 +2024,65 @@ def mock_geocoder_result(result)
+      other_student = create :student
+      Follower.create!(section_id: section.id, student_user_id: other_student.id, user: teacher)
+      assert_equal true, other_student.assigned_course_or_script?
+    end
```
Ideally, each of these would be its own test case that just sets up the conditions and does one assert.
